htsohm/htsohm.py

```diff
@@ -590,40 +617,60 @@ def worker_run_loop(run_id):
 
             if config['mutation_strength_method'] != 'flat':
-                if material.generation_index == config['children_per_generation'] - 1 and gen > 0:
-                # standard calculation of mutation strengths for all accessed bins
-                    if config['annealing_on'] != 'on' or gen % config['annealing_frequency'] != 0:
+                if config['interactive_mode'] != 'on':
+                    if material.generation_index == config['children_per_generation'] - 1 and gen > 0:
+                    # standard calculation of mutation strengths for all accessed bins
+                        if config['annealing_on'] != 'on' or gen % config['annealing_frequency'] != 0:
+                            parent_ids = get_all_parent_ids(run_id, gen)
+                            print_block('CALCULATING MUTATION STRENGTHS')
+                            ms_bins = []
+                            for parent_id in parent_ids:
+                                parent_bin = session.query(Material).get(parent_id).bin
+                                if parent_bin not in ms_bins:
+                                    print('Calculating bin-mutation-strength for bin : {}' \
+                                            .format(parent_bin))
+                                    calculate_mutation_strength(run_id, gen + 1, parent_bin)
+                                ms_bins.append(parent_material.bin)
+                        # annealing to reset all mutation strengths to initial value
+                        elif config['annealing_on'] == 'on' and gen % config['annealing_frequency'] == 0:
+                            all_accessed_bin_tuples = session \
+                                    .query(
+                                        Material.gas_adsorption_bin,
+                                        Material.surface_area_bin,
+                                        Material.void_fraction_bin) \
+                                    .filter(
+                                        Material.run_id == run_id,
+                                        Material.retest_passed != False,
+                                        Material.generation_index < config['children_per_generation']) \
+                                    .distinct().all()
+                            all_accessed_bins = [ [e[0], e[1], e[2]]
+                                    for e in all_accessed_bin_tuples]
+                            print_block('ANNEALING WITH MUTATION STRENGTH :\t{}' \
+                                    .format(config['initial_mutation_strength']))
+                            for some_bin in all_accessed_bins:
+                                print('Annealing bin :\t{}'.format(some_bin))
+                                mutation_strength = MutationStrength(run_id, gen + 1, *some_bin)
+                                mutation_strength.strength = config['annealing_strength']
+                                session.add(mutation_strength)
+
```

Comment:
The worker_run_loop method is a beast and needs to be broken up into smaller submethods...